### ISSUE TITLE
findPreviousUploads returns empty array despite the file and the uploadUrl being the same.

### ISSUE THREAD

**Issue by JulianH99** [tus/tus-js-client#238]
**Describe the bug**
I'm uploading a video with tus to vimeo and I have been asked to implement the `findPreviousUploads` functionality. But, for example, If I'm uploading a file called  `video.mp4` and close the tab, and then come back to upload the same file again, the array of previous uploads is empty. I know vimeo provides it's own uploadURL for every upload, but in this case I'm saving the uploadURL of the file in localStorage so that I can retrieve it later when the user uploads the same file again.

**Expected behavior**
The array of previous uploads should not be empty since I'm uploading the same file to the same uploadURL

**Setup details**
Please provide following details, if applicable to your situation:
- Runtime environment: Browser
- Used tus-js-client version: 2.2.0
- Used tus server software: vimeo

**Comment by Acconut** [comment]
This is not a bug but the intended behavior instead. tus-js-client will only store upload URL in localStorage if it created the upload. However, since in your case Vimeo already created the upload, and you are only passing in the `uploadUrl` option, tus-js-client will not store any URL. Since it does not store anything, `findPreviousUploads` is not able to retrieve any upload.

FYI, here is the code for storing the upload URL: https://github.com/tus/tus-js-client/blob/e145a0521f92819bb5666b7fa17532267cf1b1b3/lib/upload.js#L578

Here is the documentation stating that only created uploads will be saved: https://github.com/tus/tus-js-client/blob/master/docs/api.md#storefingerprintforresuming

As a workaround I see two options:
- Save the upload URL from Vimeo on your own in localStorage, retrieve it on your own, and then pass it to tus-js-client again when resuming.
- Retrieve the default UrlStorage and call the `addUpload` method to add the upload URL from Vimeo on your own: https://github.com/tus/tus-js-client/blob/master/docs/api.md#urlstorage

**Comment by JulianH99** [comment]
Thanks for answering. I'm currently trying the first option but still no luck, upload starts from 0 again and does not seem to be resuming anything. Any clues on what could be wrong? this is my code:
```typescript
 const uploadObject = new Upload(file, {
      uploadUrl: this.uploadUrl,
      overridePatchMethod: false,
      retryDelays: [0, 3000, 5000, 10000, 20000],
      metadata: {
        filename: file.name,
        filetype: file.type,
      },

);
```

`this.uploadUrl` is the previously returned url from the localStorage

**Comment by Acconut** [comment]
I have heard that Vimeo's tus server have a problem with resuming. Some people were able to work around this using the `chunkSize` option. If that does not help, please contact Vimeo support.

**Comment by stebogit** [comment]
Hey @Acconut I'm in the same situation and I can't really figure out how to retrieve the `UrlStorage` object.
Regardless, isn't your [second suggestion](https://github.com/tus/tus-js-client/issues/238#issuecomment-771874185) the same as calling directly `upload._saveUploadInUrlStorage();` (no need to create the entry object)? I do call that every time I pause the upload, just before calling `upload.abort()`.
Although that saves the upload object in the `localStorage`, the issue I'm having is that for some reason it keeps adding a new object every time I click on the pause button. However, if I call it only once just before starting the entire upload, it doesn't save anything in the localStorage. In both cases the upload resumes from the beginning, not from the last uploaded byte.
What's the right flow in this case?

```js
// a shortened version of my code
upload = null;
uploadIsRunning = false;

toggleBtn.addEventListener('click', function () {
    if (!upload) return;
    if (uploadIsRunning) {
        // upload._saveUploadInUrlStorage(); // this keeps adding an entry in the UrlStorage
        upload.abort();
        toggleBtn.textContent = 'resume upload';
        uploadIsRunning = false;
    } else {
        startOrResumeUpload();
    }
});

upload = new tus.Upload(file, {
    uploadUrl: uploadUrl, // from the server
    onProgress (bytesUploaded, bytesTotal) { /* update progress */ },
    onSuccess () { /* OK */ },
    onError (error) { /* handle error */ },
});

function startOrResumeUpload() {
    upload.findPreviousUploads().then(function (previousUploads) {
        if (previousUploads.length) {
            upload.resumeFromPreviousUpload(previousUploads[0])
        }
        toggleBtn.textContent = 'pause upload';
        uploadIsRunning = true;
        upload.start()
    });
}

upload._saveUploadInUrlStorage(); // this doesn't seem to work
startOrResumeUpload();
```

**Comment by Acconut** [comment]
> I can't really figure out how to retrieve the `UrlStorage` object.

It is accessible using `defaultOptions.urlStorage`. See https://github.com/tus/tus-js-client/blob/7186eebb88dc8d6b13af124f72cea66ea173253f/lib/browser/index.js#L14

> calling directly `upload._saveUploadInUrlStorage();`

Please, do not ever call private, internal methods on your own. These are not covered by any guarantees and may change at any release.

In general, I don´t understand what you are trying to do. You are already passing an `uploadUrl` to the constructor. If you do that, there is no need to retrieve the upload URL from the URL storage, because you already have it:

```js
upload = new tus.Upload(file, {
    uploadUrl: uploadUrl, // from the server
```

You can simply pause and resume the upload using `upload.abort()` and `upload.start()`. No need for `upload.findPreviousUploads()`, unless I am missing something.

**Comment by stebogit** [comment]
> Please, do not ever call private, internal methods on your own. These are not covered by any guarantees and may change at any release.

Right, that makes a lot of sense!

> You can simply pause and resume the upload using upload.abort() and upload.start(). No need for upload.findPreviousUploads(), unless I am missing something.

This does not work, as stated in the title of this issue, and in my understanding of your note it's because Vimeo requires to create the upload record via their API (that's where we get the `uploadUrl` to provide to the `tus.upload`), therefore the `tus.upload` never registers/saves the url in the `UrlStorage`

> It is accessible using defaultOptions.urlStorage

can I get it than like this, or is it not a public attribute?
```js
const storage = tus.upload.urlStorage;
```
It is likely me that is likely missing something, but I simply started out playing with the tus client by following the [provided example(s)](https://github.com/tus/tus-js-client/blob/master/docs/usage.md#example-upload-with-pause-button) with the pause option and it seems this is basically how you manage the start/stop/restart flow:

```js
upload.findPreviousUploads().then(function (previousUploads) {
  if (previousUploads.length) {
    upload.resumeFromPreviousUpload(previousUploads[0])
  }
  upload.start()
})
```
Is it not the case?
I do confirm in my case the storage always return an empty array, therefore the upload always restarts from the beginning

**Comment by Acconut** [comment]
> It is likely me that is missing something, but I simply started out playing with the tus client by following the [provided example(s)](https://github.com/tus/tus-js-client/blob/master/docs/usage.md#example-upload-with-pause-button) with the pause option and it seems this is basically how you manage the start/stop/restart flow:
> 
> ```js
> upload.findPreviousUploads().then(function (previousUploads) {
>   if (previousUploads.length) {
>     upload.resumeFromPreviousUpload(previousUploads[0])
>   }
>   upload.start()
> })
> ```
> 
> Is it not the case?

This approach is correct if you are supplying only the `endpoint` property. However, if you are using the `uploadUrl` option, `findPreviousUploads` will not have any effect. `findPreviousUploads` is used to retrieve an upload URL for a given file, but in Vimeo's case this upload URL is already known. So in your situation, it is not necessary/helpful to invoke this method at all.

To address your actual issue of Vimeo uploads not resuming from their last paused position: I doubt that this is a tus-js-client issue. Instead it sounds likely that the Vimeo server is not saving the received data correctly and therefore the upload has to start from the beginning. Please contact Vimeo support to get help from them in this regard.

> > It is accessible using defaultOptions.urlStorage
> 
> can I get it than like this, or is it not a public attribute?

No, please try `tus.defaultOptions.urlStorage`.

**Comment by stebogit** [comment]
Oh, so the last paused position is freshly fetched from the Tus server (with a HEAD or PATCH request?) when` start` is called, not saved with the `ListEntry` object in the `urlStorage`?
I'll check if there's anything wrong with the returned value/header from Vimeo. Thanks

**Comment by Acconut** [comment]
> Oh, so the last paused position is freshly fetched from the Tus server (with a HEAD or PATCH request?) when` start` is called, not saved with the `ListEntry` object in the `urlStorage`?

Yes, you are correct :) When resuming, tus-js-client will always perform a HEAD request to retrieve the current offset from the server. The data stored inside the UrlStorage do not include the file offset:

https://github.com/tus/tus-js-client/blob/7186eebb88dc8d6b13af124f72cea66ea173253f/lib/upload.js#L817-L821

Hope that clears a lot of questions up for you!
